query-engine/core/src/executor/execute_operation.rs

```diff
@@ -128,30 +127,101 @@ pub async fn execute_many_self_contained<C: Connector + Send + Sync>(
 }
 
 /// Execute the operation as a self-contained operation, if necessary wrapped in a transaction.
-async fn execute_self_contained(
+async fn execute_self_contained<'a>(
+    mut conn: Box<dyn Connection>,
+    query_schema: QuerySchemaRef,
+    operation: Operation,
+    force_transactions: bool,
+    retry_on_transient_error: bool,
+    trace_id: Option<String>,
+) -> crate::Result<ResponseData> {
+    let operation_timer = Instant::now();
+    let result = if retry_on_transient_error {
+        execute_self_contained_with_retry(
+            &mut conn,
+            query_schema,
+            operation,
+            force_transactions,
+            Instant::now(),
+            trace_id,
+        )
+        .await
+    } else {
+        let (graph, serializer) = build_graph(query_schema, operation)?;
+
+        execute_self_contained_without_retry(conn, graph, serializer, force_transactions, trace_id).await
+    };
+
+    histogram!(PRISMA_CLIENT_QUERIES_HISTOGRAM_MS, operation_timer.elapsed());
+
+    result
+}
+
+async fn execute_self_contained_without_retry(
     mut conn: Box<dyn Connection>,
     graph: QueryGraph,
     serializer: IrSerializer,
     force_transactions: bool,
     trace_id: Option<String>,
 ) -> crate::Result<ResponseData> {
-    let operation_timer = Instant::now();
-    let result = if force_transactions || graph.needs_transaction() {
-        let mut tx = conn.start_transaction(None).await?;
-        let result = execute_on(tx.as_connection_like(), graph, serializer, trace_id).await;
-
-        if result.is_ok() {
-            tx.commit().await?;
-        } else {
-            tx.rollback().await?;
+    if force_transactions || graph.needs_transaction() {
+        return execute_in_tx(&mut conn, graph, serializer, trace_id).await;
+    }
+
+    execute_on(conn.as_connection_like(), graph, serializer, trace_id).await
+}
+
+const MAX_TX_TIMEOUT_RETRY_LIMIT: Duration = Duration::from_secs(12);
```

Comment:
What was the reason to choose 12 seconds as the maximum timeout? 

My understanding based on the doc for the constant is that the errors being retried happen only when a transaction is trying to be committed but there's a network glitch or a server selection problem. 

Under those circumstances, If we loop and retry transactions for up to 12 seconds, during that period of time, many requests would pile up on the server, creating a lot of IO, and risking server availability. In addition, there's no throttling between retries, a TX is retried immediately after it has failed. For me, this signals something: Either this is intentional and the TX is retried immediately because chances are the network glitch is no longer there after a few milliseconds. Case in which we should not loop for over a few orders of magnitude more than how long it takes to likely succeed. Or the probability of a TX succeeding after a retry is equal across a window of those X seconds, case in which I think we should back-oiff before retrying.

If it is case one, I would reduce the timeout or try a fixed number of times. If it's the latter I would use exponential decay (or any other backoff strategy) over the 12 seconds window, this is: retry, if it fails again, wait for a few milliseconds, if it fails again, double the wait time, and so on until reaching the timeout. This will reduce the collective number of requests trying to reach the server upon a (supposedly) network glitch and reduce the likelihood of flooding the server.

Also depending on how frequent these errors are, and to protect server availability even further, I would consider implementing a circuit breaker. This is: if several transactions see the error repeatedly (we need shared memory) then we stop retrying on all of them and fail fast. Because the most likely thing is that the server won't recover on time, and if it does, it will get flooded but a herd of transactions is being retried. In case that made sense, https://github.com/dmexe/failsafe-rs can help with the implementation.